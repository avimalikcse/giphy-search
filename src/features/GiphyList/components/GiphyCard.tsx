import React from 'react';
import Box from '@mui/material/Box';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';
import { GIFCardDetailsType, PaginationDataType } from "../../../types/types";




type GiphyCardProps = {
    cardDetails: GIFCardDetailsType,
    setSelectedGIF: Function
}

export default function GiphyCard({ cardDetails, setSelectedGIF }: GiphyCardProps) {

    const { title, thumbnail, gifURL, height, width } = cardDetails
    return (
        <Box textAlign={'center'}>
            <Card variant="outlined" onClick={() => setSelectedGIF(cardDetails)}> <CardContent>
                <img
                    width={width}
                    height={height}
                    src={thumbnail}
                    alt={title}
                />
                <Typography variant="body2">
                    {title}
                </Typography>
                <CardActions>
                    <Button size="small">Preview</Button>
                </CardActions>
            </CardContent>

            </Card>
        </Box>
    )
}

