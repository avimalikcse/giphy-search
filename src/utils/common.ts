
/**
 * Method to prevent repeated execution of a particular function within a certain time range
 * E.G: prevent multiple API calls if use click on search button multiple time
 * 
 * @param {Function} fn function that needs to be throttled 
 * @param {number} wait time it should wait before call
 * @returns void
 */

export const throttle = (fn: any, wait: number = 300) => {
  let inThrottle: boolean,
    lastFn: ReturnType<typeof setTimeout>,
    lastTime: number;
  return function (this: any) {
    const context = this,
      args = arguments;
    if (!inThrottle) {
      fn.apply(context, args);
      lastTime = Date.now();
      inThrottle = true;
    } else {
      clearTimeout(lastFn);
      lastFn = setTimeout(() => {
        if (Date.now() - lastTime >= wait) {
          fn.apply(context, args);
          lastTime = Date.now();
        }
      }, Math.max(wait - (Date.now() - lastTime), 0));
    }
  };
};
