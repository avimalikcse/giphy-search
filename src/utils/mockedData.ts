/**
 * File to keep mocked data
 */
export const MOCKED_DATA =
{

    "id": "9JgeSP0jlRAVBOG9FD",
    "title": "Bill Nye Spinning GIF by Nike",
    "rating": "g",
    "content_url": "",
    "source_tld": "",
    "source_post_url": "",
    "is_sticker": 0,
    "import_datetime": "2018-03-02 20:36:17",
    "trending_datetime": "2018-03-05 04:20:07",
    "images": {
        "original": {
            "height": "270",
            "width": "480",
            "size": "2021711",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy.gif",
            "mp4_size": "159778",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy.mp4",
            "webp_size": "356696",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy.webp",
            "frames": "33",
            "hash": "65151deca02da1bfa8c196a328e5115b"
        },
        "downsized": {
            "height": "270",
            "width": "480",
            "size": "1110011",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-downsized.gif"
        },
        "downsized_large": {
            "height": "270",
            "width": "480",
            "size": "2021711",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy.gif"
        },
        "downsized_medium": {
            "height": "270",
            "width": "480",
            "size": "2021711",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy.gif"
        },
        "downsized_small": {
            "height": "228",
            "width": "405",
            "mp4_size": "54997",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-downsized-small.mp4"
        },
        "downsized_still": {
            "height": "270",
            "width": "480",
            "size": "34777",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-downsized_s.gif"
        },
        "fixed_height": {
            "height": "200",
            "width": "356",
            "size": "814594",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200.gif",
            "mp4_size": "88466",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200.mp4",
            "webp_size": "296408",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200.webp"
        },
        "fixed_height_downsampled": {
            "height": "200",
            "width": "356",
            "size": "165256",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200_d.gif",
            "webp_size": "100664",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200_d.webp"
        },
        "fixed_height_small": {
            "height": "100",
            "width": "178",
            "size": "269203",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100.gif",
            "mp4_size": "30509",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100.mp4",
            "webp_size": "100108",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100.webp"
        },
        "fixed_height_small_still": {
            "height": "100",
            "width": "178",
            "size": "10714",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100_s.gif"
        },
        "fixed_height_still": {
            "height": "200",
            "width": "356",
            "size": "25582",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200_s.gif"
        },
        "fixed_width": {
            "height": "113",
            "width": "200",
            "size": "327748",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200w.gif",
            "mp4_size": "36086",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200w.mp4",
            "webp_size": "121676",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200w.webp"
        },
        "fixed_width_downsampled": {
            "height": "113",
            "width": "200",
            "size": "63820",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200w_d.gif",
            "webp_size": "39244",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200w_d.webp"
        },
        "fixed_width_small": {
            "height": "57",
            "width": "100",
            "size": "105544",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100w.gif",
            "mp4_size": "13101",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100w.mp4",
            "webp_size": "42322",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100w.webp"
        },
        "fixed_width_small_still": {
            "height": "57",
            "width": "100",
            "size": "4120",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100w_s.gif"
        },
        "fixed_width_still": {
            "height": "113",
            "width": "200",
            "size": "13047",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200w_s.gif"
        },
        "looping": {
            "mp4_size": "2658625",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-loop.mp4"
        },
        "original_still": {
            "height": "270",
            "width": "480",
            "size": "90732",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy_s.gif"
        },
        "original_mp4": {
            "height": "270",
            "width": "480",
            "mp4_size": "159778",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy.mp4"
        },
        "preview": {
            "height": "122",
            "width": "216",
            "mp4_size": "19345",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-preview.mp4"
        },
        "preview_gif": {
            "height": "80",
            "width": "142",
            "size": "48869",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-preview.gif"
        },
        "preview_webp": {
            "height": "108",
            "width": "192",
            "size": "43254",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-preview.webp"
        },
        "480w_still": {
            "height": "270",
            "width": "480",
            "size": "2021711",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/480w_s.jpg"
        }
    },
    "user": {
        "avatar_url": "https://media4.giphy.com/avatars/nike/gafTSy1Qou4l.png",
        "banner_image": "https://media4.giphy.com/headers/nike/mPbbY4srw3HC.png",
        "banner_url": "https://media4.giphy.com/headers/nike/mPbbY4srw3HC.png",
        "profile_url": "https://giphy.com/nike/",
        "username": "nike",
        "display_name": "Nike",
        "description": "Just Do it",
        "instagram_url": "https://instagram.com/Nike",
        "website_url": "http://www.nike.com/justdoit",
        "is_verified": true
    },
    "analytics_response_payload": "e=Z2lmX2lkPTlKZ2VTUDBqbFJBVkJPRzlGRCZldmVudF90eXBlPUdJRl9TRUFSQ0gmY2lkPTQ2ZDdjMmI4cjg0MGx2YXRveTAzdmt0cjB3aGoyenI2dWdwczhhdXY0MXh1djJnayZjdD1n",
    "analytics": {
        "onload": {
            "url": "https://giphy-analytics.giphy.com/v2/pingback_simple?analytics_response_payload=e%3DZ2lmX2lkPTlKZ2VTUDBqbFJBVkJPRzlGRCZldmVudF90eXBlPUdJRl9TRUFSQ0gmY2lkPTQ2ZDdjMmI4cjg0MGx2YXRveTAzdmt0cjB3aGoyenI2dWdwczhhdXY0MXh1djJnayZjdD1n&action_type=SEEN"
        },
        "onclick": {
            "url": "https://giphy-analytics.giphy.com/v2/pingback_simple?analytics_response_payload=e%3DZ2lmX2lkPTlKZ2VTUDBqbFJBVkJPRzlGRCZldmVudF90eXBlPUdJRl9TRUFSQ0gmY2lkPTQ2ZDdjMmI4cjg0MGx2YXRveTAzdmt0cjB3aGoyenI2dWdwczhhdXY0MXh1djJnayZjdD1n&action_type=CLICK"
        },
        "onsent": {
            "url": "https://giphy-analytics.giphy.com/v2/pingback_simple?analytics_response_payload=e%3DZ2lmX2lkPTlKZ2VTUDBqbFJBVkJPRzlGRCZldmVudF90eXBlPUdJRl9TRUFSQ0gmY2lkPTQ2ZDdjMmI4cjg0MGx2YXRveTAzdmt0cjB3aGoyenI2dWdwczhhdXY0MXh1djJnayZjdD1n&action_type=SENT"
        }
    }
}

export const MOCKED_TYPE_DATA =
{
    id: "9JgeSP0jlRAVBOG9FD",
    title: "Bill Nye Spinning GIF by Nike",
    thumbnail: "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy_s.gif",
    gifURL: "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy_s.gif",
    width: 400,
    height: 200,
}