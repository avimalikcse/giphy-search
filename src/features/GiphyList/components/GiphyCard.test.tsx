import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { MOCKED_TYPE_DATA } from "../../../utils/mockedData";
import GiphyCard from './GiphyCard';

const onSelect = jest.fn()

function setup() {
    return render(< GiphyCard cardDetails={MOCKED_TYPE_DATA} setSelectedGIF={onSelect} />);
}

test('Test if item is displayed a card ', () => {
    setup()
    const cardTittle = screen.getByText(/Spinning/i);
    expect(cardTittle).toBeInTheDocument();
});

test('Test if item is click selecting a card', () => {
    setup()
    const cardTittle = screen.getByText(/Spinning/i);
    fireEvent.click(cardTittle)
    expect(onSelect).toBeCalled()
});


test('Test snapshots of the components', async () => {
    const { asFragment } = setup()
    await waitFor(() => {
        expect(asFragment()).toMatchSnapshot();
    })
});
