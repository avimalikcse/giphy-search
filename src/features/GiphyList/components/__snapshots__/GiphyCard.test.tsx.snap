// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshots of the components 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-xi606m"
  >
    <div
      class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
    >
       
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <img
          alt="Bill Nye Spinning GIF by Nike"
          height="200"
          src="https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy_s.gif"
          width="400"
        />
        <p
          class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
        >
          Bill Nye Spinning GIF by Nike
        </p>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
        >
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root  css-1knaqv7-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Preview
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
